packages/material-ui/src/Fab/Fab.js

```diff
@@ -1,127 +1,169 @@
+    // Reset on touch devices, it doesn't add specificity
+    '@media (hover: none)': {
+      backgroundColor: theme.palette.grey[300],
     },
+    textDecoration: 'none',
   },
-  /* Styles applied to the root element if `color="secondary"`. */
-  secondary: {
-    color: theme.palette.secondary.contrastText,
-    backgroundColor: theme.palette.secondary.main,
-    '&:hover': {
-      backgroundColor: theme.palette.secondary.dark,
-      // Reset on touch devices, it doesn't add specificity
-      '@media (hover: none)': {
-        backgroundColor: theme.palette.secondary.main,
-      },
-    },
+  '&.Mui-focusVisible': {
+    boxShadow: theme.shadows[6],
+  },
+  '&.Mui-disabled': {
+    color: theme.palette.action.disabled,
+    boxShadow: theme.shadows[0],
+    backgroundColor: theme.palette.action.disabledBackground,
   },
   /* Styles applied to the root element if `variant="extended"`. */
-  extended: {
+  ...(styleProps.variant === 'extended' && {
     borderRadius: 48 / 2,
     padding: '0 16px',
     width: 'auto',
     minHeight: 'auto',
     minWidth: 48,
     height: 48,
-    '&$sizeSmall': {
+  }),
+  ...(styleProps.variant === 'extended' &&
+    styleProps.size === 'small' && {
       width: 'auto',
       padding: '0 8px',
       borderRadius: 34 / 2,
       minWidth: 34,
       height: 34,
-    },
-    '&$sizeMedium': {
+    }),
+  ...(styleProps.variant === 'extended' &&
+    styleProps.size === 'medium' && {
       width: 'auto',
       padding: '0 16px',
       borderRadius: 40 / 2,
       minWidth: 40,
       height: 40,
-    },
-  },
-  /* Styles applied to the root element if `variant="circular"`. */
-  circular: {},
-  /* Pseudo-class applied to the ButtonBase root element if the button is keyboard focused. */
-  focusVisible: {},
-  /* Pseudo-class applied to the root element if `disabled={true}`. */
-  disabled: {},
+    }),
   /* Styles applied to the root element if `color="inherit"`. */
-  colorInherit: {
+  ...(styleProps.color === 'inherit' && {
     color: 'inherit',
-  },
+  }),
+  /* Styles applied to the root element if `color="primary"`. */
+  ...(styleProps.color === 'primary' && {
+    color: theme.palette.primary.contrastText,
+    backgroundColor: theme.palette.primary.main,
+    '&:hover': {
+      backgroundColor: theme.palette.primary.dark,
+      // Reset on touch devices, it doesn't add specificity
+      '@media (hover: none)': {
+        backgroundColor: theme.palette.primary.main,
+      },
+    },
+  }),
+  /* Styles applied to the root element if `color="secondary"`. */
+  ...(styleProps.color === 'secondary' && {
+    color: theme.palette.secondary.contrastText,
+    backgroundColor: theme.palette.secondary.main,
+    '&:hover': {
+      backgroundColor: theme.palette.secondary.dark,
+      // Reset on touch devices, it doesn't add specificity
+      '@media (hover: none)': {
+        backgroundColor: theme.palette.secondary.main,
+      },
+    },
+  }),
```

Comment:
These styles need to be in a separate callback, as the `'&:hover'` styles would replace the default ones, not augment. You can move them in the end in a separate callback. Take a look how it is done in the [Container component](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Container/Container.js#L72)

Reply:
So any time there is a `&:hover` or `&:focus` etc. Any styles that are applied when a `styleProp` condition is `true` needs to be in a separate callback?

Reply:
Exactly, mainly because if we just spread them, the second `:&hover` would completely replace the previous.